cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,920 @@
+  raft::device_span<vertex_t> edgelist_srcs,
+  raft::device_span<vertex_t> edgelist_dsts);
+
+template <typename vertex_t, typename edge_t, typename EdgeIterator>
+struct unroll_edge {
+  edge_t num_valid_edges{};
+  raft::device_span<edge_t> num_triangles{};
+  EdgeIterator edge_to_unroll_first{};
+  EdgeIterator transposed_valid_edge_first{};
+  EdgeIterator transposed_valid_edge_last{};
+  EdgeIterator transposed_invalid_edge_last{};
+
+  __device__ thrust::tuple<vertex_t, vertex_t> operator()(edge_t i) const
+  {
+    // edges are sorted with destination as key so reverse the edge when looking it
+    auto pair = thrust::make_tuple(thrust::get<1>(*(edge_to_unroll_first + i)),
+                                   thrust::get<0>(*(edge_to_unroll_first + i)));
+    // Find its position in either partition of the transposed edgelist
+    // An edge can be in found in either of the two partitions (valid or invalid)
+    auto itr = thrust::lower_bound(
+      thrust::seq, transposed_valid_edge_last, transposed_invalid_edge_last, pair);
+    size_t idx{};
+    if (itr != transposed_invalid_edge_last && *itr == pair) {
+      idx =
+        static_cast<size_t>(thrust::distance(transposed_valid_edge_last, itr) + num_valid_edges);
+    } else {
+      // The edge must be in the first boundary
+      itr = thrust::lower_bound(
+        thrust::seq, transposed_valid_edge_first, transposed_valid_edge_last, pair);
+      assert(*itr == pair);
+      idx = thrust::distance(transposed_valid_edge_first, itr);
+    }
+    cuda::atomic_ref<edge_t, cuda::thread_scope_device> atomic_counter(num_triangles[idx]);
+    auto r = atomic_counter.fetch_sub(edge_t{1}, cuda::std::memory_order_relaxed);
+  }
+};
+
+// FIXME: May re-locate this function as a general utility function for graph algorithm
+// implementations.
+template <typename vertex_t>
+rmm::device_uvector<vertex_t> compute_prefix_sum(
+  // The edgelist is segmented into 2 partitions (valid and invalid edges)
+  // and edges to be unrolled can be either in the valid or the invalid edge
+  // partition.
+  raft::handle_t const& handle,
+  raft::device_span<vertex_t const> sorted_vertices,
+  raft::device_span<vertex_t const> query_vertices)
+{
+  rmm::device_uvector<vertex_t> prefix_sum(query_vertices.size() + 1, handle.get_stream());
+
+  thrust::tabulate(
+    handle.get_thrust_policy(),
+    prefix_sum.begin(),
+    prefix_sum.begin() + query_vertices.size(),
+    [query_vertices,
+     num_edges       = sorted_vertices.size(),
+     sorted_vertices = sorted_vertices.begin()] __device__(auto idx) {
+      auto itr_lower_valid = thrust::lower_bound(
+        thrust::seq, sorted_vertices, sorted_vertices + num_edges, query_vertices[idx]);
+
+      auto itr_upper_valid = thrust::upper_bound(
+        thrust::seq, itr_lower_valid, sorted_vertices + num_edges, query_vertices[idx]);
+
+      auto dist_valid = thrust::distance(itr_lower_valid, itr_upper_valid);
+
+      return dist_valid;
```

Comment:
`itr_lower_valid`, `itr_upper_valid`, and `dist_valid` can be just `itr_lower`, `itr_lupper`, and `dist`.